### PR TITLE
fix(velodyne_decoder): vlp16 and vlp32 overflow handling

### PR DESCRIPTION
**Pull request by bgilby59**
## PR Type

- Bug Fix

## Related Links

https://github.com/tier4/nebula/pull/111

## Description

This PR extends the changes in https://github.com/tier4/nebula/pull/111. See **Nebula-only problem** in that PR description for a detailed overview of the problem.

## Review Procedure

Code review (currently do not have rosbags that can reproduce this issue).

## Pre-Review Checklist for the PR Author

**PR Author should check the checkboxes below when creating the PR.**

- [x] Assign PR to reviewer

## Checklist for the PR Reviewer

**Reviewers should check the checkboxes below before approval.**

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] (Optional) Unit tests have been written for new behavior
- [ ] PR title describes the changes

## Post-Review Checklist for the PR Author

**PR Author should check the checkboxes below before merging.**

- [ ] All open points are addressed and tracked via issues or tickets

## CI Checks

- **Build and test for PR**: Required to pass before the merge.

### CODE REVIEW

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp [hunk 1]
```diff
@@ -94,12 +94,43 @@ void Vlp16Decoder::reset_pointcloud(size_t n_pts, [[maybe_unused]] double time_s
   scan_timestamp_ = -1;
```
Comment (knzo25) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp:94]:
```suggestion
```

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp [hunk 2]
```diff
@@ -94,12 +94,43 @@ void Vlp16Decoder::reset_pointcloud(size_t n_pts, [[maybe_unused]] double time_s
...
+  const double last_overflow_time_stamp =
+    scan_timestamp_ + 1e-9 * overflow_pc_->points.back().time_stamp;
+
+  // Detect cases where there is an unacceptable time difference between the last overflow point and
+  // the first point of the next packet. In that case, there was probably a packet drop so it is
+  // better to ignore the overflow pointcloud
+  if (time_stamp - last_overflow_time_stamp > 0.05) {
+    scan_timestamp_ = -1;
+    overflow_pc_->points.clear();
+    overflow_pc_->points.reserve(max_pts_);
```
Comment (knzo25) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp:115]: Here there should be a return. I actually forgot it in my PR (I swear it was there !)

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp [hunk 3]
```diff
@@ -93,12 +93,43 @@ void Vlp32Decoder::reset_pointcloud(size_t n_pts, [[maybe_unused]] double time_s
   scan_timestamp_ = -1;
```
Comment (knzo25) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp:93]:
```suggestion
```

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp [hunk 4]
```diff
@@ -93,12 +93,43 @@ void Vlp32Decoder::reset_pointcloud(size_t n_pts, [[maybe_unused]] double time_s
...
+  const double last_overflow_time_stamp =
+    scan_timestamp_ + 1e-9 * overflow_pc_->points.back().time_stamp;
+
+  // Detect cases where there is an unacceptable time difference between the last overflow point and
+  // the first point of the next packet. In that case, there was probably a packet drop so it is
+  // better to ignore the overflow pointcloud
+  if (time_stamp - last_overflow_time_stamp > 0.05) {
+    scan_timestamp_ = -1;
+    overflow_pc_->points.clear();
+    overflow_pc_->points.reserve(max_pts_);
```
Comment (knzo25) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp:114]:
```suggestion
    overflow_pc_->points.reserve(max_pts_);
    return;
```

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp [hunk 5]
```diff
@@ -250,6 +281,9 @@ void Vlp32Decoder::unpack(const velodyne_msgs::msg::VelodynePacket & velodyne_pa
 
           intensity = raw->blocks[i].data[k + 2];
 
+          auto block_timestamp = rclcpp::Time(velodyne_packet.stamp).seconds();
```
Comment (knzo25) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp:284]: I think this one was already defined earlier in the code. Can you check?

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp [hunk 6]
```diff
@@ -230,7 +261,10 @@ void Vlp16Decoder::unpack(const velodyne_msgs::msg::VelodynePacket & velodyne_pa
                 const float z_coord = distance * sin_vert_angle;       // velodyne z
                 const uint8_t intensity = current_block.data[k + 2];
 
-                double point_time_offset = timing_offsets_[block][firing * 16 + dsr];
+                auto block_timestamp = rclcpp::Time(velodyne_packet.stamp).seconds();
```
Comment (knzo25) [inline nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp:264]: I think this one was already defined before. Can you check?

**Review by knzo25 — CHANGES_REQUESTED** [review]
Left some comments :pray:

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp (hunk 5 above)
Comment (bgilby59) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp32_decoder.cpp]: You're right! Fixed in 12077a81dd3969d7210baaf1a92440ee132bc096 along with 3 other unnecessary lines of code in the VLS128 decoder. Thanks!

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp (hunk 6 above)
Comment (bgilby59) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp]: Fixed in https://github.com/tier4/nebula/commit/12077a81dd3969d7210baaf1a92440ee132bc096

File: nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp (hunk 2 above)
Comment (bgilby59) [reply nebula_decoders/src/nebula_decoders_velodyne/decoders/vlp16_decoder.cpp:114]: Fixed in 638c532f6838975ff87846d1f11751a6233a252b

**Comment by knzo25** [comment]
@bgilby59 
Please remember to "resolve" the review comments and ask for a review again when you finish

**Comment by bgilby59** [comment]
@knzo25 Thank you for your review. I have addressed your comments. Please review the changes I made when you have time.

**Review by knzo25 — APPROVED** [review]
LGTM !

**Pull request merged**